stringdtype/stringdtype/src/casts.c

```diff
@@ -64,12 +65,293 @@ PyArrayMethod_Spec StringToStringCastSpec = {
         .slots = s2s_slots,
 };
 
+static NPY_CASTING
+unicode_to_string_resolve_descriptors(PyObject *NPY_UNUSED(self),
+                                      PyArray_DTypeMeta *NPY_UNUSED(dtypes[2]),
+                                      PyArray_Descr *given_descrs[2],
+                                      PyArray_Descr *loop_descrs[2],
+                                      npy_intp *NPY_UNUSED(view_offset))
+{
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+
+    if (given_descrs[1] == NULL) {
+        loop_descrs[1] = (PyArray_Descr *)new_stringdtype_instance();
+    }
+    else {
+        Py_INCREF(given_descrs[1]);
+        loop_descrs[1] = given_descrs[1];
+    }
+
+    return NPY_SAFE_CASTING;
+}
+
+// converts UCS4 code point to 4-byte char* assumes in is a zero-filled 4 byte
+// array returns -1 if the code point is not a valid unicode code point,
+// returns the number of bytes in the UTF-8 character on success
+static int
+ucs4_code_to_utf8_char(const Py_UCS4 code, char *c)
+{
+    if (code <= 0x7F) {
+        // 0zzzzzzz -> 0zzzzzzz
+        c[0] = (char)code;
+        return 1;
+    }
+    else if (code <= 0x07FF) {
+        // 00000yyy yyzzzzzz -> 110yyyyy 10zzzzzz
+        c[0] = (0xC0 | (code >> 6));
+        c[1] = (0x80 | (code & 0x3F));
+        return 2;
+    }
+    else if (code <= 0xFFFF) {
+        // xxxxyyyy yyzzzzzz -> 110yyyyy 10zzzzzz
+        c[0] = (0xe0 | (code >> 12));
+        c[1] = (0x80 | ((code >> 6) & 0x3f));
+        c[2] = (0x80 | (code & 0x3f));
+        return 3;
+    }
+    else if (code <= 0x10FFFF) {
+        // 00wwwxx xxxxyyyy yyzzzzzz -> 11110www 10xxxxxx 10yyyyyy 10zzzzzz
+        c[0] = (0xf0 | (code >> 18));
+        c[1] = (0x80 | ((code >> 12) & 0x3f));
+        c[2] = (0x80 | ((code >> 6) & 0x3f));
+        c[3] = (0x80 | (code & 0x3f));
+        return 4;
+    }
+    return -1;
+}
+
+static int
+unicode_to_string(PyArrayMethod_Context *context, char *const data[],
+                  npy_intp const dimensions[], npy_intp const strides[],
+                  NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyArray_Descr **descrs = context->descriptors;
+    long in_size = (descrs[0]->elsize) / 4;
+
+    npy_intp N = dimensions[0];
+    Py_UCS4 *in = (Py_UCS4 *)data[0];
+    char **out = (char **)data[1];
+
+    // 4 bytes per UCS4 character
+    npy_intp in_stride = strides[0] / 4;
+    // strides are in bytes but pointer offsets are in pointer widths, so
+    // divide by the element size (one pointer width) to get the pointer offset
+    npy_intp out_stride = strides[1] / context->descriptors[1]->elsize;
+
+    while (N--) {
+        // pessimistically allocate 4 bytes per allowed character
+        // plus one byte for the null terminator
+        char *out_buf = malloc((in_size * 4 + 1) * sizeof(char));
```

Comment:
I suspect you should strip the padding first.  The side note here is that NumPy *does* actually allow included 0s, so long there is an actual character later on! 

In the place where numpy has it, there is this little thing to find the length (different context).  Maybe a `static inline` helper?
```
    /* trim trailing zeros */
    while (ucs4len > 0 && src[ucs4len - 1] == 0) {
        ucs4len--;
    }
```

That also removes the need for `realloc` which isn't pretty...

Reply:
Thanks for the snippet, it is useful! I don't think doing this completely eliminates the need for the `realloc`, but I can avoid using it if I first strip trailing zeros in the UCS4 sequence then do an initial pass over the remaining codepoints to calculate the number of bytes needed to store the UTF8 string.

Did you have another trick in mind when you wrote this comment that would let me do the conversion with only one pass over the codepoints and one call to `malloc`?

Reply:
Ah, you are right of course. you have to do two passes two find the byte-length with unicode encoding.  Not sure what the best/typical way to do this is, I bet there are nifty tricks.  Might want to check if `PyUnicode` does anything fun when it needs to convert to utf-8?